### ISSUE TITLE
Message format device→computer

### ISSUE THREAD

**Issue by FilipDominec** [MrYsLab/Telemetrix4RpiPico#8]
1. Currently, every message type corresponds to its own reserved array in memory. Could not they share a single output buffer, possibly with a simple check to prevent collision?
2. The type of such buffers is 32-bit signed integer, however only the lowest 8 bits are transmitted (which makes transmission of actual 32-bit integers a matter of four bit-shifting commands). Could we just switch the data type to ```int8_t```, or perhaps ```uint8_t```?
3. The message array contains manually written length, which must match to the count of (manually written) trailing bytes minus one. This is redundant to the second argument in ```serial_write()```. Could this value be auto-filled? 
4. The message length is now represented by one 8-bit signed(?) integer, so a message cannot be now longer than 127 bytes (an oscilloscope or spectrometer would benefit from up to 10kB messages. Could we switch to ```uint16_t``` length?
5. I can see messaging causes short "hiccups" in real-time tasks I have tested. Have you considered putting second RP2040's core in charge of communication?

I am ready to propose some solutions and file a PR - but it would definitely cost some effort so your prior opinion is important for me. With my [discontinued  project](https://github.com/FilipDominec/rp2daq) - the features of which is about to be slowly merged into Tmx - I was able to push some 600-700 kB/s through USB. My motivation is in running instruments like electron microscope or real-time in-situ spectroscopy of semiconductor epitaxy - these can also feed the computer typically with 100s kB/s. Now I am pretty sure RP2 is capable of replacing obsoleted controller units of such machines, but cleanly and optimally implemented communication is crucial.

**Comment by FilipDominec** [comment]
6. As a separate point to be considered, couldn't we define all messages as ```struct``` (which I tend to use pervasively when coding in C) instead of flat arrays? One benefit is subjective clarity of the code, another is that it is easily to pack together different types (typically I send few ```uint8_t``` with few ```int32_t```). And finally, it should not be hard to introspect the structures from within Python code to automatically detect possible format mismatch when a new function is added. I will employ this concept in my soon-to-be-finished PR that adds stepper support − and will kindly ask you to decide.

**Comment by MrYsLab** [comment]
@FilipDominec
I will try to comment on all of your points. This comment will probably be lengthy,  so I apologize in advance.

Item #1: I am not sure what problems this solves. If you mean having a single buffer that holds multiple output messages, this may cause some issues. For example, if a message is received by the client that results in some immediate action that changes the system's state, there may be stale messages in the buffer that the client will need to discard.

Item #2 I am not sure what you are referring to. All data is converted to unsigned 8-bit values before being transmitted across the link. If both the server and client understand the data context, why does the buffer type matter?

Item #3 I am not sure if I understand how to determine the message length without having an end of message delimiter added to the message.

Item #4 The design philosophy of Telemetrix is to send short messages across the link to ensure a minimal amount of blocking. To send a large data block, one would break the stream into a series of messages and then reassemble on the client-side. 

Item #5 Using separate cores adds a complexity that I am unwilling to tackle.

Item #6 Defining things as structs on the server side makes sense, but interpreting structs on the client side is somewhat painful. 

My suggestion would be to fork Telemetrix and create a separate project to meet your specific needs. Incorporating your suggestions into the Telemetrix repository would result in breaking changes for existing users. I am in the middle of several other projects, so my time is somewhat limited, but I would be happy to answer any questions you might have.

**Comment by FilipDominec** [comment]
Alan, thanks for your answer. I, too, apologize for getting a into a bit broader context here, but my motivation has to be clarified. Raspberry Pi Pico seems like a new and almost perfect solution for most of laboratory automation, and the concept of preparing one universal firmware and defining its behaviour through Python is very sound. 

It is thus no surprise that Telemetrix and rp2daq share this concept and emerged roughly at same time. Some other people may have started similar projects or soon they will, I guess. I am not against choice and healthy competition in general. Having more programs to choose is fine when we talk about web browsers or games etc. − but in such a specific case of open-sourced hardware interface, fragmentation instead brings much confusion (for potential users) and risk of duplicated work (for developers). 

As a truly terrible counter example, I can point to the discussion [about defragmenting Python-based instrumentation control](https://github.com/python-data-acquisition/meta/issues/14). About 10−12 years ago, several teams independently started writing libraries that facilitate interaction with lab instruments. This is also an exciting idea that can boost scientific productivity greatly, and it nicely complements with using RP2 for some less demanding tasks. Alas, none of these projects got popular enough, so other teams found it obviously easier to start off their own projects (c.f. https://xkcd.com/927/). Currently [there are about 38](https://ethercalc.net/1anmq248ktu6) partially duplicated projects. I can confirm it is very hard to choose one, and even harder is to make it work correctly as they mostly suffer from lack of developer attention. 

In short, I had started a separate project, realized I would bear responsibility for potential fragmentation, and decided to discontinue it in favor of collaborating on Tmx, having in mind the tao of Python ```There should be one-- and preferably only one --obvious way to do it.```

**Comment by FilipDominec** [comment]
Technically,
1. unless the buffers eat up too much RAM, it is just a matter of programming style. But long messages >10kB may need a different approach. (And they *will* be needed for some tasks.)
2. With ```uint_8t``` buffers, it is quite easy e.g. to use idiomatic ```memcpy(out_buf+2, &timestamp[m], sizeof(timestamp[m]));``` rather than [manual byte-per-byte copy](https://github.com/MrYsLab/Telemetrix4RpiPico/blob/master/Telemetrix4RpiPico.c#L371). AFAIK this breaks nothing, and compiler will catch any type mismatch.
3. ```serial_write()``` can auto-fill the byte (or better, word) once it is called with a given length
4. This is a good point; single threaded server may (or may not) lag if long message is to be transmitted. But <128byte messages may bring significant overhead if a lot of data is to be transferred. I will have to find out the optimum.
5. I, too, have no experience with multi-threading on RP2. If only outgoing transfers are to be handled, it may be quite easy, though. I should try it. 
6. With ```import struct```, the Pythonic definitions can go quite similar to the C code. In my opinion, it brings nice symmetry across the two languages (so mis-matches are easy to notice).

I would like to avoid fragmentation. My time is scarce, and I have already postponed some experiments to make them finally run with Tmx.  I can maintain a "progressive" fork of Tmx from which selected features can be pulled back if needed by somebody else. But changes that break communication (like support for 128+ byte messages) seems not like one of them. I will have to think what to do next.

**Comment by MrYsLab** [comment]
At this point in time, I do not foresee any additional features that I will be adding for Rpi Pico, so it is probably a low risk for you to fork and start a new project that you have control over. An alternative is for me to create a branch, but since my time, too, is limited, this may not be ideal.
Since your needs are very specific to your application, going separate routes may satisfy both of our needs without making time-constrained demands on each other. Of course, if I can, I will supply support to aid in your endeavor.

**Comment by FilipDominec** [comment]
Apparently our needs and visions are diverging and it may be a wise decision for me to work on a separate project. I will generally aim to try to squeeze out maximum performance from RP2, and this stipulates substantial changes that break communication compatibility (also addressing #11 as side effect). 

Therefore, simplest solution for me will be to revive [rp2daq](https://github.com/FilipDominec/rp2daq) I have previously put on hold. The threaded asynchronous communication with computer is however a clever concept I will build on.

**Comment by MrYsLab** [comment]
Sounds like a plan. Thanks for your input.
